### ISSUE TITLE
"Interface" in interface names

### ISSUE THREAD

**Issue by mindplay-dk** [php-fig/fig-standards#442]
Having words like "Interface" in interface names is generally considered pretty poor practice, like having the word "Class" in class-names, "object" in variable-names, etc.

I've actually already seen people aliasing these interfaces in code, e.g.
```php
use Psr\Http\Message\ResponseInterface as Response
```
so that their code comes out more semantic - my function accepts a "request" or a "response", it definitely does not accept a "request interface" as an argument, because an "interface" is something abstract and not something you can pass as an argument at all.

In the case of this standard, the abstract concept being defined is that of a "request" or a "response", not of an "interface", which is a concept defined by the programming language.

I don't know, I think it's pretty bad for a community standard to propagate a bad practice like this from the get-go... It doesn't make us look very good to the rest of the development world.

**Comment by henriquemoody** [comment]
I agree with you in some aspects but if I'm not wrong you should probably send an email to the mailing list instead creating an issue here.
Have you read [002-psr-naming-conventions.md](https://github.com/php-fig/fig-standards/blob/master/bylaws/002-psr-naming-conventions.md))? It's not a PSR spec but it's an standard for FIG in itself. I didn't follow the discussion about this but I guess if it is in the fig-standards repository, they probably had many discussions before. Just my 2 cents.

**Comment by mindplay-dk** [comment]
Whoa, it goes further than that, huh. Wow, that's pretty messed up. For consistency, they really should have a convention to suffix class-names with "Class" ;-)

**Comment by drgomesp** [comment]
They should also prefix functions with `function`:

```function functionDoSomething() {}```

After all, it's much more readable. :facepunch:

**Comment by mindplay-dk** [comment]
The only argument I can think of in favor of this nonsense is avoiding namespace collisions - but, the interfaces/traits are already deeply namespaced, so... makes no real sense to me.

**Comment by milesj** [comment]
+1

I tend to see a lot of implementations where the abstract class uses the non-suffixed/prefixed name. For example, the interface is `ContainerInterface`, and the abstract class is `Container`. This is just weird to me as the abstract class is still that, abstract. You're neither type hinting on it either.

**Comment by mindplay-dk** [comment]
I always feel insecure about using things with language terminology artifacts in the names. Why? because it makes me think someone was struggling to come up with an appropriate, descriptive name - which suggests the thing they were implementing or describing wasn't well-defined to begin with, or have overlapping concerns.

I realize that there are no overlapping concerns (or other problems) and it doesn't indicate a problem in this case - but it does inspire and tempt developers to create/derive classes/abstractions with nearly identical names (sans suffix) and likely with overlapping concerns...

**Comment by graste** [comment]
Apart from the naming conventions psr I can see one advantage of the `Interface` suffix: When browsing code I didn't write myself it's easier to spot the contracts of the library. Easier to spot in file listings and easier to understand in source code as well. I don't need to second guess whether method parameter type hints are concrete implementations or a contract of some sort. That are pro `Interface` suffix arguments for me. Maybe I'm old.

**Comment by mindplay-dk** [comment]
> Easier to spot in file listings and easier to understand in source code as well. I don't need to second guess whether method parameter type hints are concrete implementations or a contract of some sort.

I get that. The problem is, extending the same line of argumentation, you could argue that classes are clearer with a "Class" suffix, that naming an integer variable with an "i" prefix makes it easier to spot integers, and so forth.

IMO, better not to start down that road.

Instead, learn to use an IDE, where information already present in the source code is exposed to you - rather than manually codifying redundant information into your symbol names.

IMO such practices are remnants of a time when we didn't have proper tools. That doesn't make you "old" - solid tools for PHP have only been around for probably 5-6 years or so; but they are here now and we don't need to mangle our code or pollute it with information about the code, which is already present in the code :-)

**Comment by weierophinney** [comment]
We debated this heavily in the ZF community at one point. There were two compelling arguments that were made, and which swayed those voting on our standards:

- For non-native-English speakers, descriptive naming becomes difficult, both when creating interfaces as well as identifying artifacts that might be interfaces. When creating a new interface, the `-able` suffix is quite often a natural fit, but for many words, there's no easy way to add such a suffix, and finding adequate verbiage is hard for native speakers, and often next-to-impossible for those who are not. In terms of identifying what is and is not an interface, you then need to have excellent English knowledge in order to identify "descriptive" names vs concrete (noun) names.
- As @graste noted, quite a number of developers argued that they want to be able to quickly scan file listings and `use` statements to identify what code is a contract, vs what code is not. Yes, tools can help identify these things, but there's also a very strong argument to be made that you should not _require_ tools in order to understand and use code. Your argument that the "practices are remnants of a time when we didn't have proper tools" makes a rather large assumption: that everybody gravitates to the same tools you do. Not everybody uses, or wants to use, an IDE in order to code.

Whether or not you find these reasons compelling isn't terribly relevant to me; the fact that they were the arguments that influenced the community I work with is, as it's not just a handful of people making the decision, but a majority of users backing it. I suspect similar sentiments were present when FIG adopted these naming guidelines.

**Comment by pmjones** [comment]
Hear hear.

**Comment by donquixote** [comment]
> The problem is, extending the same line of argumentation, you could argue that classes are clearer with a "Class" suffix

I would say that this argument is bogus.
There are classes, interfaces and traits. Class is the "default thing" (e.g. in a directory listing), interfaces and traits (and possibly abstract classes) are the non-default things. So it is fine if interfaces and traits get a suffix, but classes don't.

To distinguish items from n categories, it is sufficient to leave one category without label, and label only the remaining n-1 categories.

**Comment by donquixote** [comment]
For the same reason it would be valid to label only private and protected methods, and omit the "public" keyword on public methods, because this is the default.
We chose (somewhere, I think) that "public" should not be omitted. But it would have been a valid choice to say it should be omitted.

And, saying that one out of n is the "default" does not even mean it is the majority case. It is simply an agreement.

**Comment by mindplay-dk** [comment]
> There are classes, interfaces and traits. Class is the "default thing"

Well, I would say that's bogus.

The only place a class name is the default thing, is in a `new` statement, and in the `extends` clause of a `class` declaration - and that's no different from how interfaces are the default thing in an `implements` clause, traits are the default thing in a `use` clause, and by "default", in all of these cases, of course I mean the only possible thing.

When it comes to type-hints in function arguments and return type-hints (with the exception of traits, which can't be type-checked) the type of "thing" doesn't even matter - interfaces and classes are completely equal, since the only thing you're really type-hinting against in either case, is the public interface of that "thing".

The only counter-argument that would make any sense to me is "we're already doing that" - I'm suprised nobody threw that in the mix. PHP codebases are already littered with this stuff, and I'm doing it myself now, because, standards, consistency, and "all the other kids were doing it"...

Bottom line, it doesn't matter if you're doing it right or wrong - if you break rank, it looks wrong.

So I'll resign myself to doing it wrong with the rest of the PHP community, and we can all feel right doing it wrong, because apparently enough people doing it wrong makes it right ;-)

Also: whatevs.

**Comment by donquixote** [comment]
Well.. there could be a parallel universe where "Interface" is the "default thing". And by "default" I mean nothing but convention and familiarity. The one of the types/things that was lucky, and does not need a prefix or suffix.

E.g. you have 4 teams, which are distinguishable by their socks: One with blue socks, one with red socks, one with green socks, and one without socks. This way we save 1/4 of the cost for buying the socks. Or possibly more, if the teams are different in number. Which of the teams is sockless does not matter, as long as everyone agrees. So yes, "we're already doing that".

And unlike humans, these items do not have a right to equal treatment, or to have their feet covered. Appending *Interface does not force us to also append *Class. And appending *Class would not force us to append *Interface.

And btw I would rather append *ConcreteImplementation, not *Class, because the goal is (imo) to describe the role of the element, not to repeat the modifier keyword 1:1. But since, by tradition, "concrete implementation" is the lucky one, I leave it without a suffix.

**Comment by weierophinney** [comment]
I was going to sit this one out, but there were two interesting points made earlier I want to comment on, both made by @mindplay-dk :

> When it comes to type-hints in function arguments and return type-hints (with the exception of traits, which can't be type-checked) the type of "thing" doesn't even matter - interfaces and classes are completely equal, since the only thing you're really type-hinting against in either case, is the public interface of that "thing".

and:

> The only counter-argument that would make any sense to me is "we're already doing that" - I'm suprised nobody threw that in the mix. PHP codebases are already littered with this stuff,

I'm going  to turn this around, and ask you to ask: "why is everybody doing that?" Having witnessed these debates for more than a decade, I have some perspective to share.

Regarding the first argument, it's a matter of context. When looking inside the code itself, the context is typically apparent with `use`, `implements`, and `extends`. In the case of typehints and `instanceof` checks, it becomes murkier. As Rasmus notes, all types are equal in those regards — but on the flip side, when I'm looking at the code, I often want to know if the typehint is for a *contract* vs an *implementation*, as it gives me additional context. Implementations raise alarms, as it means I'm coding against specific behavior vs a general contract, and I'll often start looking at refactoring:

- does an interface exist already? if so, why am I not typehinting against it?
- if an interface exists, can I extract one, so that I can have substitutions?

Since the typehints cannot tell me what I'm looking at, that means I have to do some digging in order to determine that information. When `Interface` is directly in the name, I know already I'm dealing with a contract.

Another context to consider is when I'm browsing a directory of files. If the filename does not indicate if the file is a trait or an interface, I'm left with a bunch of code that all looks equal. Quite often, I'm looking for *contracts*, so I know what to code against, or what to implement. When all files look equal, I have to use tools to determine what files to look into. Similarly, I often want to see what traits are available to accomplish common implementation aspects of interfaces; again, if all files look equal, this is harder to accomplish.

Yes, I know, IDEs can simplify this. Not everyone can or chooses to use IDEs, however; for some, cost or availability is a factor; for others, accessibility; for others, ideology. My point is that the *contexts* in which code is examined are not singular, and should be considered as well.

Another context to consider is when an interface and an implementation sit side-by-side in the same namespace. Quite often, the default implementation may not have an identifying characteristic that would allow it to have an easily separated name. As an example: I might want to define "routes", and the primary implementation is rather generic — it doesn't pull from a database, or a configuration file, etc, but simply implements the methods in the contract. The inclination is to have a `Route` interface and a …. `Route` class? How can I differentiate? (Please, don't answer; it's a rhetorical question. I already can provide rationale for either approach.) `DefaultRoute` or `GenericRoute` or `BaseRoute` are horrible names, but all too common when using this approach.

This leads to my final point, which addresses both quoted sections from above.

In the Zend Framework community, we had a lengthy and healthy debate about naming interfaces, abstract classes, and traits quite some years ago. One thing that's always been interesting to me about the ZF community is how *international* it is; I'd argue that native English speakers are a small minority of ZF's users *and* developers. Why does that matter? Because many of our core developers argued that they had difficulty coming up with a good interface names, and identifying existing interface names *due to the language barrier*. As such, an overwhelming majority backed the idea of suffixing interfaces and traits, and prefixing abstract classes; doing so simplified the contribution process by simplifying one of the hardest jobs in programming: naming things.

Yes, it may be ugly and distasteful. Yes, there may be [good practices for naming](http://mnapoli.fr/approaching-coding-style-rationally/) to follow, but:

- Not all domains lend themselves equally to such naming, making consistency hard to achieve.
- Not all developers use the same tools, making identification of contracts difficult.

Those reasons *in some cases* trump the ideal approach. Projects that have large numbers of developers speaking many different tongues will have different needs than small, homogenous projects, and will therefor need to make different choices, some of which may fall outside an ideal.

With that said, I think adoption of the such naming practices *should be done on a per-project* approach. FIG has put it into the by-laws *so that all interfaces it defines follow consistent naming*. It is *not* however, part of PSR-1/2, as, again, it can and should be determined by the project.
